src/net_processing.cpp

```diff
@@ -3988,11 +3988,56 @@ void PeerLogicValidation::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerLogicValidation::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
-    // Check whether we have too many outbound peers
-    int extra_peers = m_connman.GetExtraOutboundCount();
+    // If we have any extra block-relay peers, disconnect the youngest unless
+    // it's given us a block -- in which case, disconnect the second youngest
+    // instead (so we rotate our youngest block-relay peer).
+    int extra_block_relay_peers = m_connman.GetExtraBlockRelayCount();
+    if (extra_block_relay_peers > 0) {
+        // Pick the block-relay peer that we just connected, unless it gave us
+        // a block -- in which case disconnect the next-youngest block-relay
+        // peer.
+        std::pair<NodeId, int64_t> youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+        std::pair<NodeId, int64_t> second_youngest_peer = {-1, std::numeric_limits<int64_t>::max()};
+
+        m_connman.ForEachNode([&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            if (!pnode->IsBlockOnlyConn() || pnode->fDisconnect) return;
+            if (pnode->GetId() > youngest_peer.first) {
+                second_youngest_peer = youngest_peer;
+                youngest_peer.first = pnode->GetId();
+                youngest_peer.second = pnode->nLastBlockTime;
```

Comment:
This is a bit of a name collusion, maybe `second_youngest_peer` -> `next_youngest_peer`

Reply:
Changed.